pkg/devfile/generator/generators.go

```diff
@@ -91,6 +98,70 @@ func GetContainers(devfileObj parser.DevfileObj, options common.DevfileOptions)
 	return containers, nil
 }
 
+// GetInitContainers gets the init container for every preStart devfile event
+func GetInitContainers(devfileObj parser.DevfileObj) ([]corev1.Container, error) {
+	containers, err := GetContainers(devfileObj, common.DevfileOptions{})
+	if err != nil {
+		return nil, err
+	}
+	preStartEvents := devfileObj.Data.GetEvents().PreStart
+	var initContainers []corev1.Container
+	if len(preStartEvents) > 0 {
+		var eventCommands []string
+		commands, err := devfileObj.Data.GetCommands(common.DevfileOptions{})
+		if err != nil {
+			return nil, err
+		}
+
+		commandsMap := getCommandsMap(commands)
+
+		for _, event := range preStartEvents {
+			eventSubCommands := getCommandsFromEvent(commandsMap, strings.ToLower(event))
+			eventCommands = append(eventCommands, eventSubCommands...)
+		}
+
+		for i, commandName := range eventCommands {
+			if command, ok := commandsMap[commandName]; ok {
+				component := common.GetExecComponent(command)
+				commandLine := common.GetExecCommandLine(command)
+				workingDir := common.GetExecWorkingDir(command)
```

Comment:
so i think this is now invalid which is why odo [disabled](https://github.com/openshift/odo/blob/main/pkg/devfile/validate/events.go#L5) it as per my comment prev because `preStart` needs to be `apply`/`composite`(of only `apply` subcommands) https://github.com/devfile/api/blob/main/pkg/validation/validation-rule.md#events

Reply:
updated